File: JavaScript Interview Code/142. Linked List Cycle II.js



var detectCycle = function(head) {
    let fast = head;
    let slow = head;

    // here we need to make sure the fast and fast.next value
    while (fast.next != null && fast.next.next != null) {
        fast = fast.next.next;
        slow = slow.next;
        if (fast == slow) {
            slow = head;
            // find the starting point;
            while (fast != slow) {
                fast = fast.next;
                slow = slow.next;
            }
            return slow;
        }
    }
    return null;
}

File: Daily Coding Practice/30 -day finish 300 coding challenge/2020-4-27. Fast and Slow Pointer pattern/142. Linked List Cycle II.py
class Solution:
    def detectCycle(self, head):
        if not head: return
        # calculate if we have cycle linked list
        k = self.catch_up(head)
        if k == False: return None

        fast = head
        slow = head
        while k:
            fast = fast.next
            k -= 1
        
        while fast != slow:
            fast = fast.next
            slow = slow.next
        return fast
    
    def num_cycle(self, head):
        cur = head.next
        count = 1
        while cur != head:
            cur = cur.next
            count += 1
        return count
    
    def catch_up(self, head):
        fast = head
        slow = head

        while fast.next and fast.next.next:
            fast = fast.next.next
            slow =slow.next
            if fast == slow:
                return num_cycle(fast)
        return False

# time complexity o(N)
# space complexity o(1)

